予約機能の実装

fryota142 opened the pull request:
FPユーザがカレンダーから予約枠を作成して、userが予約枠を取れるようにした。
reservationのmodel, view, controllerを追加した。

fryota142 committed:
create reservation model.

予約のためのモデルを追加した

fryota142 committed:
add reservation controller.

fryota142 committed:
add fullcalender.

カレンダーを追加した

fryota142 committed:
implement reservation function for user

fp_userが予約枠を作成できるようにした
userが予約枠から予約を取れるようにした

fryota142 committed:
ユーザ情報に予約一覧を表示するようにした

fryota142 committed:
FPユーザ情報の予約枠一覧を表示する

fryota142 committed:
予約情報の見た目を変更した

fryota142 committed:
fix fullcalendar function.

カレンダーが動作しない問題を解決した
カレンダーコントローラーの追加

haw-fukata committed:
implement edit fp_user.

FPユーザに編集機能を追加した
view_showのユーザ情報を見やすくした

haw-fukata committed:
implement edit user.

ユーザ情報の編集を行えるようにした
ユーザ情報を見やすくした

haw-fukata committed:
add validate start_time of reservation.

予約機能に曜日ごとの時間帯によるバリデーションを追加した
予約の削除機能を追加した

haw-fukata committed:
add image and partially changed of header and top.

fryota142 committed:
add reservation list to calendar.

カレンダーに予約情報を表示した

fryota142 committed:
create reservation function on dayclick

カレンダーの日付をクリックして予約枠を追加できるようにした

fryota142 committed:
add calendar appointment for user.

ユーザがカレンダーにある予約から詳細ページに飛んで予約できるようにした

fryota142 committed:
remove unnecessary file.

不要なファイルを削除した

eguchi-ken commented on app/helpers/calendars_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module CalendarsHelper
+end
```
使わなさそうなファイルは検索のときにじゃまになったりするので、消しておくほうがよいですね。

eguchi-ken commented on app/controllers/fp_users_controller.rb:
```diff
@@ -16,13 +16,27 @@ def create
 
   def show
     @fp_user = FpUser.find(params[:id])
+    @reservations = Reservation.all.where(fp_user_id: current_user.id)
```
このアクションは、下のどちらの用途で使うものでしょうか？

(A) FP が自分の予約を見るためのもの
(B) 普通のユーザーが FP の予約状況をみるためのもの

それによって current_user.id を使うかどうかが変わってくると思います。もし (A) の目的で用意している画面なら id は受け取らず 18行目は下のようにするのがよいと思います。

```diff
- @fp_user = FpUser.find(params[:id])
+ @fp_user = current_user
```

(B) の目的で用意している画面なら、その FP の予約状況を見るためのものなので下のようにするべきです。

```suggestion
    @reservations = Reservation.all.where(fp_user_id: @fp_user.id)
```

また、他のユーザーの予約が見えるべきではない、というふうに考えるのであれば下のようになるでしょう。

```suggestion
    @reservations = Reservation.all.where(fp_user_id: @fp_user.id, user_id: current_user.id)
```

fryota142 replied:
(B)の方で考えていました。
今見返すと、よくわからない中途半端な抽出になってしまっていました。
レビューありがとうございます！

eguchi-ken commented on app/views/fp_users/edit.html.slim:
```diff
@@ -0,0 +1,14 @@
+.col-md-8.mx-auto
+  = link_to "戻る", :back, class: "btn btn-outline-success"
```
できれば URL helper を使って、どこに画面遷移させたいかを明示したほうが良いと思います。

https://railsguides.jp/routing.html#%E3%83%91%E3%82%B9%E3%81%A8url%E7%94%A8%E3%83%98%E3%83%AB%E3%83%91%E3%83%BC

:back だとブラウザバックになりHTTPリファラを参照するので、期待した画面に戻せる保証がないからです。

fryota142 replied:
参考までありがとうございます！
変更します

eguchi-ken commented on app/views/fp_users/edit.html.slim:
```diff
@@ -0,0 +1,14 @@
+  .col-md-6.mx-auto
+    = bootstrap_form_for @fp_user do |f|
+      = f.text_field :name, label: "名前"
+      = f.text_area :introduction, label: "自己紹介"
```
メールアドレスやパスワードの変更もできるようにしたほうがよいですね。

eguchi-ken commented on app/views/fp_users/show.html.slim:
```diff
@@ -1,3 +1,36 @@
+    br
+    = "自己紹介"
+    br
+    = "#{@fp_user.introduction}"
```
文字列との結合がないので、式展開もしなくてよいと思います。

```suggestion
    = @fp_user.introduction
```

eguchi-ken commented on app/views/fp_users/show.html.slim:
```diff
@@ -1,3 +1,36 @@
+br
+.text-center
+  h2 予約枠一覧
+- if @reservations.present?
```
予約がないときに何も表示されないと、ユーザーが不安になるので「予約がありません」といったメッセージや、予約するための画面へ進むリンクがあるとよいかと思います。

fryota142 replied:
確かにこのままだと不安なので、userのshowも一緒にelseにメッセージとリンクを追加しました。
レビューありがとうございます!

eguchi-ken commented on app/views/fp_users/show.html.slim:
```diff
@@ -1,3 +1,36 @@
+  table.table.table-hover.table-sm.table-responsive-sm.text-nowrap
+    thead
+      tr
+        th User
```
ここだけ英語になっているのちょっと気になります。

fryota142 replied:
「ユーザ名」に変更しました

eguchi-ken commented on app/controllers/fp_users_controller.rb:
```diff
@@ -16,13 +16,27 @@ def create
+
+  def update
+    @fp_user = FpUser.find(params[:id])
+    if @fp_user.update(fpuser_params)
```
この処理は、current_user が誰であっても実行できるようになっていないでしょうか？
もしそうだとすると、勝手に名前を書き換えたり攻撃されてしまうことになります。更新しようとしている対象が current_user と一致しているか、などは見ておいたほうがよいと思います。同じことは予約の削除などについてもいえます。

その助けになる gem もあります。いつか使うかもしれませんが、あまり時間がないので、今回は使わなくて良いと思います。

fryota142 replied:
privateにcorrect_userを追加して、一致しているか確認するようにしました
ありがとうございます！

eguchi-ken commented on app/models/user.rb:
```diff
@@ -2,4 +2,6 @@ class User < ApplicationRecord
   validates :email, presence: true, uniqueness: true, length: { maximum: 64 }
   has_secure_password
+
+  has_one :reservation
```
user は複数の予約を持つことができるはずなので、 has_many であるべきだと思います。

eguchi-ken commented on app/models/reservation.rb:
```diff
@@ -0,0 +1,18 @@
+class Reservation < ApplicationRecord
```
同じ人が、同じ日、同じ時間に予約してしまうダブルブッキングが発生することはないですか？

fryota142 replied:
validateにcheck_double_bookingを追加して対応しました
ありがとうございます!

eguchi-ken commented on app/models/fp_user.rb:
```diff
@@ -2,4 +2,6 @@ class FpUser < ApplicationRecord
   validates :email, presence: true, uniqueness: true, length: { maximum: 64 }
   has_secure_password
+
+  has_one :reservation
```
FP は複数の予約を持つことができるはずなので、 has_many であるべきだと思います。

eguchi-ken reviewed (commented):
@fryota142 レビューしました。

haw-fukata committed:
implement calendar for user and fp_user.

ユーザとFPで別のカレンダーを表示するようにした。
FPは時間をクリックした際に予約枠が作成できるようにしており、ユーザは予約枠の確保だけできるようにした

haw-fukata committed:
add validation and edit to reservation.

ユーザがダブルブッキングをしないようにvalidationを追加した
showでのボタンにキャンセル等を追加した
edit画面を追加した

haw-fukata committed:
refactoring user and fp_user. change time zone.

ユーザとFP新規登録時に自動でログインするようにした
変更などを加える際に本人かどうかの確認を追加した
modelのhas_oneをhas_manyに変更した
timezoneを日本に変更した

haw-fukata committed:
refactoring appearance.

ユーザとFPのshowをレスポンシブに対応した
予約で使っているtableの文字の変更を行った
headerの予約のpathをcalendarに飛ぶようにした

fryota142 committed:
fix show of user and fp.

FPとユーザの曖昧な抽出条件の修正
reservationのcorrect_fp_userのidの変更を行った

fryota142 committed:
implement update reservation for fp_user

FPユーザが予約枠の時間を変更できるようにした

fryota142 committed:
add flash message and readme.

flashのメッセージを日本語に変更した
crud時にflashでメッセージが出るようにした
readmeの内容を変更した

fryota142 committed:
add show user and fpuser from reservation.

予約詳細画面からuserとfpuserの情報を見れるようにした
